### PR TITLE
Macro upgrades techs others

### PR DESCRIPTION
**Pull request by syhw**
I tested the lua and python examples, and dump_replays.lua.

Have yet to test state.hasResearched(BW::TechType) and state.getUpgradeLevel(BW::UpgradeType).

### CODE REVIEW

File: quick_setup.sh [hunk 1]
```diff
@@ -21,10 +21,10 @@ cp $TC_DIR/BWEnv/bin/*.dll ./
 cp $TC_DIR/BWEnv/bin/*.sh ./
 cp $TC_DIR/config/*.ini $SC_DIR/bwapi-data/
 cp $TC_DIR/maps/micro/ $SC_DIR/Maps/BroodWar/ -r
-curl -LO https://github.com/TorchCraft/TorchCraft/releases/download/v1.3-0/torchcraft-v1.3-0.zip
-unzip torchcraft-v1.3-0.zip
-cp torchcraft-v1.3-0/* ./
-cp torchcraft-v1.3-0/bin/* ./
+curl -LO https://github.com/TorchCraft/TorchCraft/releases/download/v1.3-1/torchcraft-v1.3-1.zip
+unzip torchcraft-v1.3-1.zip
```
Comment (ebetica) [inline quick_setup.sh]: Can we do this in a different pr? After my frame diff and creep map change

File: include/frame.h [hunk 2]
```diff
@@ -141,11 +151,163 @@ struct Unit {
...
+    Argus_Jewel             = 1ll << 47,
+    Argus_Talisman          = 1ll << 49,
+    Caduceus_Reactor        = 1ll << 51,
+    Chitinous_Plating       = 1ll << 52,
+    Anabolic_Synthesis      = 1ll << 53,
+    Charon_Boosters         = 1ll << 54,
+    Upgrade_60              = 1ll << 60,
+    Unknow                  = 1ll << 62,
+  };
+  enum UpgradesLevel : uint64_t { // could be uint32_t
```
Comment (ebetica) [inline include/frame.h]: Put the level 1 upgrades in here too?

**Comment by ebetica** [comment]
This is also missing frame diff + diff serialization support.

File: include/frame.h (hunk 2 above)
Comment (syhw) [inline include/frame.h]: They're already in Upgrades

File: BWEnv/src/controller.cc [hunk 3]
```diff
@@ -1007,6 +1030,15 @@ void Controller::addUnit(BWAPI::Unit u, replayer::Frame& frame, BWAPI::PlayerInt
...
     u->getResources(),
+    u->getBuildType().getID(),
+    u->getTech().getID(),
+    u->getUpgrade().getID(),
+    u->getRemainingBuildTime() + u->getRemainingTrainTime(),
+    u->getRemainingResearchTime() + u->getRemainingUpgradeTime(),
+    u->getSpellCooldown(),
+    associatedUnit,
+    u->getScarabCount() + u->getSpiderMineCount() + u->getInterceptorCount(),
+    u->hasNuke(),
```
Comment (syhw) [inline BWEnv/src/controller.cc]: You mean in the counts? ok

File: BWEnv/src/controller.cc [hunk 4]
```diff
@@ -926,6 +943,12 @@ void Controller::addUnit(BWAPI::Unit u, replayer::Frame& frame, BWAPI::PlayerInt
     }
   }
 
+  int32_t associatedUnit = -1;
+  if (u->getAddon() != nullptr) associatedUnit = u->getAddon()->getID();
+  else if (u->getTransport() != nullptr) associatedUnit = u->getTransport()->getID();
+  else if (u->getHatchery() != nullptr) associatedUnit = u->getHatchery()->getID();
+  else if (u->getNydusExit() != nullptr) associatedUnit = u->getNydusExit()->getID();
```
Comment (syhw) [inline BWEnv/src/controller.cc]: That's given by getNydusExit(), it's symmetric.

File: quick_setup.sh (hunk 1 above)
Comment (syhw) [inline quick_setup.sh]: ok

File: include/frame.h (hunk 2 above)
Comment (ebetica) [inline include/frame.h]: I meant move them into UpgradesLevel, it makes more sense to me this way to have all the level upgrades in a single enum.

File: py/pyconstants.cpp [hunk 5]
```diff
@@ -56,6 +56,7 @@ void init_constants(py::module& torchcraft) {
   setEnumDict<BW::UserCommandType>(constants, "usercommandtypes");
   setEnumDict<BW::Order>(constants, "orders");
   setEnumDict<BW::TechType>(constants, "techtypes");
+  setEnumDict<BW::UpgradeType>(constants, "upgradetypes");
```
Comment (ebetica) [inline py/pyconstants.cpp]: I think this should be Upgrades, UpgradesLevel, and Techs, not UpgradeType.

Comment (syhw) [inline py/pyconstants.cpp]: I think your referring to the state values, that's in py/pyreplayer.cpp. These are the constants/enums.

File: include/frame.h (hunk 2 above)
Comment (syhw) [inline include/frame.h]: These enums are local to resources, I do not think anybody should use the UpgradesLevel enum, without the getUpgradeLevel(BW::UpgradeType) helper.
